shared/middleware/helpers/bucket_helper.rb

```diff
@@ -125,9 +126,25 @@ def copy_files(src_channel, dest_channel, options={})
 
       src = "#{@bucket}/#{src_prefix}#{filename}"
       dest = s3_path dest_owner_id, dest_channel_id, filename
-      response = s3.copy_object(bucket: @bucket, key: dest, copy_source: URI.encode(src), metadata_directive: 'REPLACE')
 
-      {filename: filename, category: category, size: fileinfo.size, versionId: response.version_id}
+      begin
+        response = s3.copy_object(bucket: @bucket, key: dest, copy_source: URI.encode(src), metadata_directive: 'REPLACE')
+        {filename: filename, category: category, size: fileinfo.size, versionId: response.version_id}
+      rescue Aws::S3::Errors::InvalidArgument
+        # Capturing additional information to help diagnose this error where we
+        # occasionally pass a bad copy_source to the S3 API.
+        # https://app.honeybadger.io/projects/3240/faults/35329035/8aba7532-c087-11e7-8280-13b5745130ae
+        # Rescuing and re-raising lets us add additional information while
+        # implicitly re-wrapping the original exception.
+        # http://blog.honeybadger.io/nested-errors-in-ruby-with-exception-cause/
+        raise CDOError.new <<-ERR
+S3 object copy failed with the following parameters:
+  bucket: #{@bucket}
+  key: #{dest}
+  copy_source: #{URI.encode(src)}
+  metadata_directive: 'REPLACE'
+        ERR
+      end
```

Comment:
I think you could also use `Honeybadger.context` to [add context to Honeybadger errors](http://docs.honeybadger.io/ruby/getting-started/adding-context-to-errors.html), if that would be useful.

Reply:
Oh, that's way better. Thanks!